packages/flutter_tools/lib/src/ios/devices.dart

```diff
@@ -893,16 +944,24 @@ class IOSDeviceLogReader extends DeviceLogReader {
   // Strip off the logging metadata (leave the category), or just echo the line.
   String _debuggerLineHandler(String line) => _debuggerLoggingRegex.firstMatch(line)?.group(1) ?? line;
 
+  /// Use both logs from `idevicesyslog` and `ios-deploy` when debugging from CI system
+  /// since sometimes `ios-deploy` does not return the device logs:
+  /// https://github.com/flutter/flutter/issues/121231
+  @visibleForTesting
+  bool get useBothLogDeviceReaders {
+    return _usingCISystem && _majorSdkVersion >= 16;
+  }
+
   void _listenToSysLog() {
     // syslog is not written on iOS 13+.
-    if (_majorSdkVersion >= minimumUniversalLoggingSdkVersion) {
+    if (!useBothLogDeviceReaders && _majorSdkVersion >= minimumUniversalLoggingSdkVersion) {
       return;
     }
     _iMobileDevice.startLogger(_deviceId).then<void>((Process process) {
       process.stdout.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).listen(_newSyslogLineHandler());
       process.stderr.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).listen(_newSyslogLineHandler());
       process.exitCode.whenComplete(() {
-        if (linesController.hasListener) {
+        if (!(useBothLogDeviceReaders && debuggerStream != null) && linesController.hasListener) {
```

Comment:
could you expand this condition into multiple expressions, as I'm getting confused trying to read it. Maybe:

```dart
if (!linesController.hasListener) {
  return;
}
if (useBothLogDeviceReaders && debuggerStream != null) {
  return;
}
linesController.close();
```